opal/datatype/opal_datatype_internal.h

```diff
@@ -156,8 +156,8 @@ typedef struct ddt_elem_id_description ddt_elem_id_description;
  */
 struct ddt_elem_desc {
     ddt_elem_id_description common;           /**< basic data description and flags */
-    uint32_t                blocklen;         /**< number of elements on each block */
-    size_t                  count;            /**< number of blocks */
+    size_t                  blocklen;         /**< number of elements on each block */
+    uint32_t                count;            /**< number of blocks */
```

Comment:
Why change `count` from `size_t` to `uint32_t` here? It's being cast to `size_t` in many places now, which seems error prone. It seems safer to handle everything as `size_t` inside the implementation.

Reply:
As the PR log indicates the count mimics what we receive from MPI, right now an int32_t. The only protection we need is to prevent the optimizer from merging elements if their summed count is larger than uint32_t.

Allowing both fields to be size_t would have changed the size of the structure, and this structure is critical to be as small as possible (and cache aligned).

The cast added in many places is to ensure that the compiler generates the correct coda, aka promote the result to size_t.

Reply:
In that case: your change changes the size of this struct ;) By putting the 64bit `size_t` after the 2x16-bit `ddt_elem_id_description` member the compiler is required to insert padding (4B each) before `blocklen` and after `count` (before `extent`, which again is 64-bit). Swapping the two members would fix this, of course.

AFAICS, the struct was previously 48B, now it's 56B. Making both members size_t would have the same effect and the struct would still fit into one cache line.

Reply:
Good catch, I forgot to swap the 2 fields. The size of the ddt_elem_desc_t struct is 32B (aka 4 * 8B) on 64 bits architectures.

I am not sure what would be the point of having the count larger than uint32_t, but if you have a valid use case let me know.

Reply:
Ahh right, it is 32bit. Not sure I how got to 48...

> I am not sure what would be the point of having the count larger than uint32_t, but if you have a valid use case let me know.

My rational was that having the internal handling in `size_t` across the board (i.e., everything behind the MPI interface) helps avoid future overflows, e.g., because of a forgotten cast somewhere. Plus, I assume this structure will have to be adapted anyway if/when BigCount lands, right?

Reply:
Maybe. It does not makes sense to have both the blocklen and the count as size_t, as together that would represent a gigantic memory span (2^128).